Added color to the output of the embUnit "standard" and of the "text" outputters

noshky opened the pull request:
This change prints succesful test runs in GREEN, and failed runs in RED.

The failed message was slightly adjusted to a more meaningful message:
Before: run 10 failures 3
Now: FAILED (3 of 10 failed)

(It has to be checked that no external tools rely on this output message. If that would be the case, this change should be discarded.)

authmillenon commented:
There are some terminal software who do not support color output. So I would strongly advice against hard-coded color input into RIOT. Though we mainly use `pyterm` for our work, there is no reason for using other terminal software, like e.g. PuTTY on Windows (bad example since it supports color, but I guess you get my point).

authmillenon commented:
As for the other output changes: It might be helpful for new devs ;-).

noshky commented:
Good point, i think Eclipses console may be a better example. It does color, but i've seen strange output with standard terminal color codes before.

Speaking of which... for me, unit-testing really mainly is about seeing green or red. So i'm very interested in having color output. What do you think about leaving the default embUnit outputters alone and adding a ``ColorTextOutputter``?

authmillenon commented:
Maybe a better approach, since the outputter can be chosen at compile time (for those who don't know: with environment variable `OUTPUT`; the options are `TEXT`,  `XML`,  `COMPILER` or anything for default behavior). Maybe a `ColorOutputter` for default output with colors would also be possible. XML and compiler(only output on failure) would not benefit much from colorful output.

noshky commented:
Ok i have already added this outputtet.

Since I'm not too familar with GIT and PRs, could you advise how i would include this to this PR? Can I rebase (undo) the first change that changes the default behaviour, and add this new outputter? Or should I create a new PR?

noshky commented:
So i've forced-push the new ColorTextOutputter now, and undone the changes that I've made to the standard outputters.
Now we only have the verbose outputter in color (the one that prints a line for every single testcase). Do we want another outputter for the short summary, like the default non-color-outputter?

authmillenon commented:
Yes please :-)

authmillenon commented:
(but add it as a separate commit please)

noshky commented:
There it is. We now have
``OUTPUT=COLOR`` for simple color output (non-verbose), according to ``OUTPUT=``(nothing).
and
``OUTPUT=COLORTEXT`` for detailed output in color, according to ``OUTPUT=TEXT``.

authmillenon commented on sys/embunit/ColorOutputter.c:
```diff
@@ -0,0 +1,102 @@
+{
+    (void) self;
+    (void) test;
+    (void) runCount;
```
Shouldn't there be a `.` in here somewhere?

authmillenon commented on sys/embunit/ColorOutputter.c:
```diff
@@ -0,0 +1,102 @@
+    (void) msg;
+    (void) line;
+    (void) file;
+    (void) runCount;
```
The standard outputter prints the test case and the line of code on failure. Also the failed assertion.

authmillenon commented on sys/embunit/ColorOutputter.c:
```diff
@@ -0,0 +1,102 @@
+{
+    (void) self;
+    if (result->failureCount) {
+        printf("\n\033[41;1mFAILED\033[21m (%d of %d failed)", result->failureCount,
```
I think for readability it's better to replace the color escape sequences with some macros here (and in the other outputter):

```C
#define CRED   "\033[41;1m"
#define CGREEN "\033[42;1m"
```

And then use string concatenation to form the format string:

```C
printf("\n" CRED "FAILED" ...);
```

noshky commented:
Thanks for your feedback!
I've added the "." so that the output is now similar to the default output (but green :-)). Also, I replaced the color codes with macros.

authmillenon commented on sys/include/embUnit/ColorOutputter.h:
```diff
@@ -0,0 +1,71 @@
+#define BGDEFAULT   "\033[49m"
+#define SBOLD       "\033[1m"
+#define SDEFAULT    "\033[21m"
+#define LINEFILL    "\033[K"
```
I think it's better if those stay private (in the scope of the respective C files).

authmillenon commented on sys/embunit/ColorOutputter.c:
```diff
@@ -0,0 +1,100 @@
+    (void) self;
+    (void) runCount;
+    printf("\n" CRED "FAILED %s (%s:%d) %s" CDEFAULT "\n", Test_name(test), file, line, msg);
+}
```
Just had a look at the original implementation: https://github.com/RIOT-OS/RIOT/blob/master/sys/embunit/TestRunner.c#L46 maybe it's better to not use printf, since it is notoriously big. 

Also I was mistaken: the `.` is printed every time.

noshky commented:
I made the color definitions private by moving them to a header file. Didn't want duplicated definitions in both source files.

To the ``printf``, well the default implementation also resolves to ``printf``. Unless someone wanted to implement a ``stdimpl_print`` function, changing the outputter to not use printf wouldn't make much of a difference. Also, all the other outputters use printf.

authmillenon commented on sys/include/embUnit/ColorOutputter.h:
```diff
@@ -0,0 +1,52 @@
+ * authorization of the copyright holder.
+ *
+ * $Id: ColorOutputter.h,v 1.2 2003/09/06 13:28:27 arms22 Exp $
+ */
```
Please use the standard LGPL license header we use in all our other contributions. This file was not written by the Embedded Unit Project in 2003, but in 2015 by you ;-).

authmillenon commented on sys/include/embUnit/ColorTextOutputter.h:
```diff
@@ -0,0 +1,50 @@
+ * authorization of the copyright holder.
+ *
+ * $Id: ColorTextOutputter.h,v 1.2 2003/09/06 13:28:27 arms22 Exp $
+ */
```
Same as above.

authmillenon replied:
Still wrong license? Or by choice?

authmillenon commented on sys/embunit/ColorTextOutputter.c:
```diff
@@ -0,0 +1,90 @@
+ * authorization of the copyright holder.
+ *
+ * $Id: ColorTextOutputter.c,v 1.4 2003/09/06 13:28:27 arms22 Exp $
+ */
```
License

authmillenon commented on sys/embunit/ColorTextColors.h:
```diff
@@ -0,0 +1,65 @@
+ * authorization of the copyright holder.
+ *
+ * $Id: ColorTextOutputter.h,v 1.2 2003/09/06 13:28:27 arms22 Exp $
+ */
```
License

authmillenon commented on sys/embunit/ColorOutputter.c:
```diff
@@ -0,0 +1,101 @@
+ * authorization of the copyright holder.
+ *
+ * $Id: ColorOutputter.c,v 1.4 2003/09/06 13:28:27 arms22 Exp $
+ */
```
License

authmillenon commented:
Apart from the license issue, this PR looks good now.

authmillenon commented:
Please notify me if you addressed my comments so I can get to it faster ;-). Apart from the license in `sys/include/embUnit/ColorTextOutputter.h` everything looks (still) good :-)

noshky commented:
Oh sorry for this, will keep you updated :-).
I've added the missing license piece and done some squashing. Also I have written a word to the color-outputters into ``/tests/unittests/README.md``. Is this the right/only place for it? Does it need more explanation? Btw, I found the existing documentation on unittests really helpful, it got me started to testing real easy.

Should I squash the commits down to just one?

authmillenon commented:
First of all: Please squash a PR only if it was ACK'd. This way I can easier track the changes you've made. The "NEEDS SQUASHING" label is primarily a warning for us maintainers to not merge temporary commits in ;-)

authmillenon commented:
ACK and yes, please squash the commits down to one commit.

authmillenon commented:
@noshky ping?

authmillenon commented:
Still not squashed

OlegHahm commented:
Superseeded by #3639.
